File: examples/triton_emitter/kernels_py/vector_exp.py
"""Triton vector-exp + backward — direct copy of tritonbench's reference kernels.

The original carries an optional `profile_mem` pointer hooked to a
`tritonbench.kernels.profile.time()` helper; that's a benchmarking concern, not
part of the kernel's math. We drop it here so the IR comparison is purely the
forward/backward computation.
"""
import triton
import triton.language as tl


@triton.jit
def triton_exp_kernel(
    x_ptr,
    output_ptr,
    n_elements,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    x = tl.load(x_ptr + offsets, mask=mask)
    output = tl.exp(x)
    tl.store(output_ptr + offsets, output, mask=mask)


@triton.jit
def triton_exp_backward_kernel(
    grad_output_ptr,
    output_ptr,
    grad_input_ptr,
    n_elements,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    grad_output = tl.load(grad_output_ptr + offsets, mask=mask)
    output = tl.load(output_ptr + offsets, mask=mask)
    grad_input = grad_output * output
    tl.store(grad_input_ptr + offsets, grad_input, mask=mask)

File: examples/triton_emitter/kernels_zig/vector_exp.zig
//! Zig ports of `kernels_py/vector_exp.py` (`triton_exp_kernel` +
//! `triton_exp_backward_kernel`).

const tri = @import("zml").kernel.triton;
const zml = @import("zml");

pub const VectorExpFwd = struct {
    pub const Cfg = struct { BLOCK_SIZE: i32 = 1024 };
    pub const Kernel = tri.Kernel(Cfg, .{
        .name = "triton_exp_kernel",
        .inputs = &.{ "x_ptr", "n_elements" },
        .outputs = &.{"output"},
        .run = run,
    });
    fn run(b: *tri.Builder, cfg: Cfg) tri.FinishError!void {
        const a = try b.declareArgs(.{
            .x_ptr = .{ .ptr = .f32 },
            .output_ptr = .{ .ptr = .f32 },
            .n_elements = .{ .scalar_opts = .{ .dtype = .i32, .divisibility = 16 } },
        });

        const pid = b.programId(.x);
        const offsets = pid.mul(cfg.BLOCK_SIZE).add(b.arange(0, cfg.BLOCK_SIZE, .i32));
        const mask = offsets.lt(a.n_elements);

        const x = b.loadOpts(a.x_ptr.addPtr(offsets), .{ .mask = mask });
        const out = b.exp(x);
        b.storeOpts(a.output_ptr.addPtr(offsets), out, .{ .mask = mask });
    }
};

pub const VectorExpBwd = struct {
    pub const Cfg = struct { BLOCK_SIZE: i32 = 1024 };
    pub const Kernel = tri.Kernel(Cfg, .{
        .name = "triton_exp_backward_kernel",
        .inputs = &.{ "grad_output_ptr", "output_ptr", "n_elements" },
        .outputs = &.{"grad_input"},
        .run = run,
    });
    fn run(b: *tri.Builder, cfg: Cfg) tri.FinishError!void {
        const a = try b.declareArgs(.{
            .grad_output_ptr = .{ .ptr = .f32 },
            .output_ptr = .{ .ptr = .f32 },
            .grad_input_ptr = .{ .ptr = .f32 },
            .n_elements = .{ .scalar_opts = .{ .dtype = .i32, .divisibility = 16 } },
        });

        const pid = b.programId(.x);
        const offsets = pid.mul(cfg.BLOCK_SIZE).add(b.arange(0, cfg.BLOCK_SIZE, .i32));
        const mask = offsets.lt(a.n_elements);

        const grad_output = b.loadOpts(a.grad_output_ptr.addPtr(offsets), .{ .mask = mask });
        const output = b.loadOpts(a.output_ptr.addPtr(offsets), .{ .mask = mask });
        const grad_input = grad_output.mul(output);
        b.storeOpts(a.grad_input_ptr.addPtr(offsets), grad_input, .{ .mask = mask });
    }
};
